### PR TITLE
rustdoc: various cross-crate reexport fixes

### PR DESCRIPTION
**Pull request by fmease**
Fixes for various smaller cross-crate reexport issues.
The PR is split into several commits for easier review. Will be squashed after approval.

Most notable changes:

* We finally render late-bound lifetimes in the generic parameter list of cross-crate functions & methods.
  Previously, we would display the re-export of `pub fn f<'s>(x: &'s str) {}` as `pub fn f(x: &'s str)`
* We now render unnamed parameters of cross-crate functions and function pointers as underscores
  since that's exactly what we do for local definitions, too. Mentioned as a bug in #44306.
* From now on, the rendering of cross-crate trait-object types is more correct:
  * `for<>` parameter lists (for higher-ranked lifetimes) are now shown
  * the return type of `Fn{,Mut,Once}` trait bounds is now displayed

Regarding the last list item, here is a diff for visualization (before vs. after):

```patch
- dyn FnOnce(&'any str) + 'static
+ dyn for<'any> FnOnce(&'any str) -> bool + 'static
```

The redundant `+ 'static` will be removed in a follow-up PR that will hide trait-object lifetime-bounds if they coincide with [their default](https://doc.rust-lang.org/reference/lifetime-elision.html#default-trait-object-lifetimes) (see [Zulip discussion](https://rust-lang.zulipchat.com/#narrow/stream/266220-rustdoc/topic/clean_middle_ty.3A.20I.20need.20to.20add.20a.20parameter/near/307143097)). `FIXME(fmease)`s were added.

@rustbot label A-cross-crate-reexports
r? @GuillaumeGomez

### CODE REVIEW

File: src/librustdoc/clean/mod.rs
Comment (cjgillot) [inline src/librustdoc/clean/mod.rs:1170]: Why is this required? The output of `fn_sig` should already be deterministic.

File: src/librustdoc/clean/mod.rs
Comment (cjgillot) [inline src/librustdoc/clean/mod.rs:1159]: Please do not reference that PR. Nobody knows *if* it will be merged at all.

File: src/librustdoc/clean/mod.rs
Comment (fmease) [reply src/librustdoc/clean/mod.rs:1170]: Maybe *stable* & *deterministic* is not the correct choice of words here. I mostly mean *predictable*.
I am sure `fn_sig` is deterministic but I am more worried about `collect_referenced_late_bound_regions` which returns a `FxHashSet` and which I think is the culprit:

Assuming I don't sort the late-bound lifetimes alphabetically, then their order in following code snippet depends on the shape of the rest of the code. If I comment out the methods `T::a`–`T::l`, then the parameter list of `T::test` is rendered as `<'a, 'b, 'c>` (on my machine at least). If I get rid of the comment again, then it is rendered as `<'c, 'a, 'b>` (in this case). Changing the name of `T::test` (e.g. to `sep`) also changes the order around.

<details>
<summary>code</summary>

```rust
pub struct T;
impl T {
    pub fn a<'x>(_: fn(&'x ()) -> &'x ()) {}
    pub fn b<'x>(_: fn(&'x ()) -> &'x ()) {}
    pub fn c<'x>(_: fn(&'x ()) -> &'x ()) {}
    pub fn d<'x>(_: fn(&'x ()) -> &'x ()) {}
    pub fn e<'x>(_: fn(&'x ()) -> &'x ()) {}
    pub fn f<'x>(_: fn(&'x ()) -> &'x ()) {}
    pub fn g<'x>(_: fn(&'x ()) -> &'x ()) {}
    pub fn h<'x>(_: fn(&'x ()) -> &'x ()) {}
    pub fn i<'x>(_: fn(&'x ()) -> &'x ()) {}
    pub fn j<'x>(_: fn(&'x ()) -> &'x ()) {}
    pub fn k<'x>(_: fn(&'x ()) -> &'x ()) {}
    pub fn l<'x>(_: fn(&'x ()) -> &'x ()) {}

    pub fn test<'a, 'b, 'c>(_: &'a &'b str, _: &(dyn ToString + 'c)) {}
}
```

</details>

Comment (fmease) [reply src/librustdoc/clean/mod.rs:1170]: I am going to update all those comments to use the word *predictable* instead and to mention the hash set.

**Review by GuillaumeGomez — APPROVED** [review]

**Comment by GuillaumeGomez** [comment]
Looks good to me, thanks for working on this. Let's wait for @cjgillot's approval too. Also I'd like to have @notriddle's confirmation too.

File: src/librustdoc/clean/inline.rs
Comment (cjgillot) [inline src/librustdoc/clean/inline.rs:263]: You should just use `sig.bound_vars()`, which gives the kinds and names of the late-bound variables, and avoid any unpredictability. This will avoid sorting, since they are in source code order.

File: src/librustdoc/clean/mod.rs
Comment (cjgillot) [inline src/librustdoc/clean/mod.rs:1161]: Likewise.

File: src/librustdoc/clean/inline.rs
Comment (fmease) [reply src/librustdoc/clean/inline.rs:263]: Ah, that's helpful! Thanks :)

File: src/librustdoc/clean/mod.rs
Comment (cjgillot) [inline src/librustdoc/clean/mod.rs:1706]: There is an object lifetime default query, so you can just test equality.

File: src/librustdoc/clean/mod.rs
Comment (cjgillot) [inline src/librustdoc/clean/mod.rs:1736]: Concatenating + deduplicating pb.bound_vars() should do the right thing here.

File: src/librustdoc/clean/mod.rs
Comment (fmease) [reply src/librustdoc/clean/mod.rs:1706]: Yep, [I am aware](https://rust-lang.zulipchat.com/#narrow/stream/266220-rustdoc/topic/clean_middle_ty.3A.20I.20need.20to.20add.20a.20parameter/near/307143097) of the existence of `tcx.object_lifetime_default` but it's not the whole solution to the problem, it's only part of it.

The query doesn't just take a lifetime, it takes the `DefId` of a type parameter of an item, e.g. the `T` of `Box`. Anyways, that's all fine and good and explained in the linked Zulip topic. I'd really prefer to implement that in a follow-up PR to avoid bitrotting and to make reviewing easier.

File: src/librustdoc/clean/mod.rs
Comment (fmease) [reply src/librustdoc/clean/mod.rs:1736]: Ye, I thought about using `bound_vars` here as well after you recommended it to me. However, I'd need to collect the iterator into an `FxIndexSet` for deduplication and immediately after convert it into a `Vec` (since that's what the rest of the code expects) which requires reallocation I assume. That's why I dropped the idea but maybe it's not too bad actually.

File: src/librustdoc/clean/mod.rs
Comment (fmease) [reply src/librustdoc/clean/mod.rs:1706]: Right, so regarding trait-object-type lifetime-bound defaults, my local work-in-progress patch is currently +168/-43 in size and it's not as straightforward as this PR meaning I won't add it to it.

**Comment by cjgillot** [comment]
@GuillaumeGomez LGTM on the use of rustc data structures.

**Comment by GuillaumeGomez** [comment]
Thanks to both of you!

@bors r=cjgillot,GuillaumeGomez rollup

**Pull request merged**
